[SU-229][SU-231] New file browser part 2: directory tree

nawatts opened the pull request:
This is part 2 ([part 1](https://github.com/DataBiosphere/terra-ui/pull/3468)) of adding a new file browser to replace the files section in the workspace Data tab.

This adds a directory tree and a very basic files list (improved in [part 3](https://github.com/DataBiosphere/terra-ui/pull/3487)). Following PRs will add features like uploading and deleting files, etc.

To access the new file browser, enable the "Workspace Files Browser" feature preview (from http://localhost:3000/#feature-preview) and use the files button in the right sidebar.

![Screen Shot 2022-11-03 at 11 01 37 AM](https://user-images.githubusercontent.com/1156625/199759543-53dc439c-2834-4cd8-8dbd-a846941ff312.png)

To try pagination, add a small `pageSize` parameter to the GCSFileBrowserBackend in WorkspaceFiles.ts. For example,
```ts
GCSFileBrowserBackend({ bucket: bucketName, project: googleProject, pageSize: 5 })
```

nawatts committed:
Add workspace files page

nawatts committed:
Add directory tree

cahrens commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,172 @@
+      ]))
+    ]),
+    status === 'Loading' && renderDirectoryStatus(level, 'Loading...'),
+    hasNextPage && h(Link, {
```
Note that Links will become Buttons if they don't have an href defined.

nawatts replied:
I know, but Link seems to be our component for a text only button.

cahrens replied:
Clickable doesn't work? I ask because this caused confusion on a recent PR of ours-- the author assumed that creating a Link would result in an anchor tag.

nawatts replied:
That would lose the visual styles on Link.
https://github.com/DataBiosphere/terra-ui/blob/c31123e6328f56789203ba3a8f31e60786e61446/src/components/common.js#L95-L100

cahrens replied:
Thanks, I now think I have to go back and fix something. :)

cahrens commented:
@nawatts I'd like to take a look at this from an a11y perspective when it's ready and you have a deployment up.

cahrens commented on src/pages/workspaces/workspace/analysis/ContextBar.js:
```diff
@@ -218,7 +218,16 @@ export const ContextBar = ({
           ...Utils.newTabLinkProps
-        }, [icon('terminal', { size: 40 }), span({ className: 'sr-only' }, ['Terminal button'])])
+        }, [icon('terminal', { size: 40 }), span({ className: 'sr-only' }, ['Terminal button'])]),
+        h(Clickable, {
```
This is enabled for Azure workspaces, but results in an error.

cahrens replied:
Also, should this be behind a feature flag?

nawatts replied:
Oops, didn't consider Azure workspaces. And yes, I forgot to add the feature flag.

cahrens replied:
Remember to disable on Azure as well as the feature flag… and let me know if you figure out how to make the keyboard navigation more natural. These roles likely would help: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role.

nawatts replied:
Thanks for the reminder. I'm working on updating this to use tree roles and add keyboard nav.

cahrens commented:
Should this be within a "Files" tab like the other Workspace "Views"? It feels like it's disconnected.

![image](https://user-images.githubusercontent.com/484484/199800688-7b3da191-dc26-42c8-990d-8b23550f8e94.png)

vs.

![image](https://user-images.githubusercontent.com/484484/199800759-4ab1c71d-83bf-465d-9a39-23276c994187.png)

cahrens commented:
I'm having trouble interacting with the tree view via the keyboard-- expand and collapsing elements. I think it's because your focus can move into the list item element, at which point toggling doesn't do anything.

I would expect to be able to expand collapse tree items with the arrow keys (like how it works in Finder), and move up and down with arrow keys. I don't know if we have any place in Terra that behaves in this way though.

cahrens commented:
This role will probably help with getting the more natural keyboard interactions: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role

nawatts commented:
> Should this be within a "Files" tab like the other Workspace "Views"? It feels like it's disconnected.

The button in the sidebar vs a tab was UX's decision. I've been thinking of it as the sidebar views are for more direct access to cloud resources while the tabs are Terra abstractions.

aherbst-broad commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,172 @@
+
+
+interface FileBrowserDirectoryContentsProps {
+  provider: FileBrowserProvider
```
I like that we're abstracting the file browser behind this interface 👍

aherbst-broad commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,172 @@
+    loadNextPage
+  } = useDirectoriesInDirectory(provider, path)
+
+  if (status === 'Error') {
```
I'm curious, do we get enums with typescript, and if so would that be something to leverage for these sorts of comparisons?

nawatts replied:
TypeScript does have [enums](https://www.typescriptlang.org/docs/handbook/enums.html), but we've elected to prefer [unions of literals](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types) over them. For example:
```ts
const Status = 'Loading' | 'Ready' | 'Error'

if (status === 'Loading') { ... }
```
vs
```ts
enum Status {
  Loading,
  Ready,
  Error,
}

if (status === Status.Loading) { ... }
```

See https://www.bam.tech/article/should-you-use-enums-or-union-types-in-typescript for more information on the pros and cons of each.

We do get type checking here. For example, if I change this line to `if (status === 'error')` then TypeScript complains:
> This condition will always return 'false' since the types '"Loading" | "Ready" | "Error"' and '"error"' have no overlap.

aherbst-broad reviewed (approved):
I played with this locally, looking good.

MatthewBemis reviewed (approved):
Tested locally, code looks good. :+1: pending putting this behind a feature flag

nawatts committed:
Improve a11y of directory tree

nawatts commented:
Updated the directory tree to use the tree role. Tested with VoiceOver in Chrome and Safari. It's not perfect, but much better than it was. Still need to add the feature flag and remove the link for Azure workspaces.

nawatts committed:
Hide workspace files link for Azure workspaces

nawatts committed:
Add feature flag

cahrens commented:
Thanks @nawatts, this was a lot of work, and the way the left and right keys work to expand/collapse items definitely feels more natural now! I did hit this issue though after I collapsed the tree:

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/484484/201747645-ac892298-8362-463f-a4cd-468dd515ea5e.png">

Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?). It actually makes me wonder if there is some 3rd party library that we should be using for an accessible tree view.

cahrens commented:
The duplication of "Files" feels odd here.

<img width="312" alt="image" src="https://user-images.githubusercontent.com/484484/201748408-46074dec-8f15-45f8-9105-3e6e6f32a855.png">

cahrens commented:
@nawatts did you happen to look at https://dgreene1.github.io/react-accessible-treeview/? It's the first thing that came up when I searched… don't know what else is out there.

nawatts committed:
Show spinner while loading directory contents

nawatts committed:
Add label for root directory

nawatts committed:
Add screen reader announcements for loading/loaded directory contents

nawatts committed:
Update tests

nawatts commented:
> Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?).

Added screen reader announcements for loading/loaded based on https://dgreene1.github.io/react-accessible-treeview/docs/examples-MultiSelectCheckboxAsync. Also changed the visual loading indicator to a spinner instead of a "Loading" message.

As for the right panel, I was thinking of moving focus there when a directory was selected.

> The duplication of "Files" feels odd here.

Needed some label for the root path (''). Replaced "Files" with "Workspace bucket".

> I did hit this issue though after I collapsed the tree:

Where are you seeing that warning? I'm only seeing complaints about color contrast.

cahrens commented:
To see the axe error (I'm using Axe development tools on your PR site), expand some of the nodes that load content, then collapse them. It is always reproducible for me.

<img width="1477" alt="image" src="https://user-images.githubusercontent.com/484484/201990263-3c0f15c3-ad21-4e24-9ede-6b00cf659e09.png">

nawatts commented:
Ah, found it. It's occurs when the tree item with keyboard focus is in the collapsed subtree. Then the aria-activedescendant attribute on the tree refers to an element that doesn't exist.

nawatts committed:
Update aria-activedescendant when subtree is collapsed

cahrens commented:
I really like the aria live regions for loading state! And I like the spinner much more than the "Loading" subitem-- thanks.

The only wish I still have is that for some reason the screen reader is not consistently announcing things as you try to navigate back out of child directories (including collapsing items), so it can be somewhat confusing to figure out where you are. But possibly screen reader folks know special ways of interacting with trees? Not sure if there's something actionable here for you.

I'll start taking a look at the code this afternoon, between meetings.

nawatts commented:
Yeah, I found that to be the case even on some of the library demos, such as https://dgreene1.github.io/react-accessible-treeview/docs/examples-DirectoryTree.

nawatts committed:
Add tooltip for link in context bar

nawatts commented:
> I feel like UX needs to sort out how navigation between these views is supposed to work. The mix of breadcrumbs, tabs, and the context bar is pretty confusing… it doesn't seem clear how to get back to the tabbed view once you have gone into files. Who is the UX person you are working with?

@joyliu10

kyuksel commented:
> > Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?).
> 
> Added screen reader announcements for loading/loaded based on https://dgreene1.github.io/react-accessible-treeview/docs/examples-MultiSelectCheckboxAsync. Also changed the visual loading indicator to a spinner instead of a "Loading" message.

Does the PR address [SUP-777](https://broadworkbench.atlassian.net/browse/SUP-777) too then?

cahrens commented:
Does the PR address [SUP-777](https://broadworkbench.atlassian.net/browse/SUP-777) too then?

No, @kyuksel. This PR is just about the tree view for workspace data (loading when a node is expanded). SUP-777 is about our general loading mechanism.

cahrens commented on src/pages/workspaces/workspace/WorkspaceFiles.ts:
```diff
@@ -0,0 +1,45 @@
+
+
+export const WorkspaceFiles = _.flow(
+  forwardRefWithName('WorkspaceFiles'),
```
I know we use `forwardRefWithName` all over the place, but I can't figure out what it actually does (and there is no documentation). Do you happen to know?

nawatts replied:
Since we use anonymous arrow functions, React can't figure out a component name for something like:
```js
const Component = forwardRef((props, ref) => ...)
```

`forwardRefWithName` adds a name to the component function so that it is identifiable in the Components panel of the React dev tools.

These would have the same effect:
```js
const Component = forwardRefWithName('MyComponent', (props, ref) => ...)
```

```js
const Component = forwardRef(function MyComponent(props, ref) { ... })
```

nawatts replied:
As for why it's needed here (despite `ref` being unused), WorkspaceContainer attaches a ref to its wrapped component. Without `forwardRef` here, React throws a warning: "Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?"

https://github.com/DataBiosphere/terra-ui/blob/448e871bde8e429ed1092b27969b5ee97d8df34c/src/pages/workspaces/workspace/WorkspaceContainer.js#L417

cahrens commented on src/pages/workspaces/workspace/WorkspaceFiles.ts:
```diff
@@ -0,0 +1,45 @@
+import _ from 'lodash/fp'
```
Can you name this file just "Files.ts"? The workspace part is already well implied by the directories, and it will help with CODEOWNERS as the workspaces team won't get flagged on changes to this file.

nawatts committed:
Rename module

nawatts committed:
Show tab bar

cahrens commented on src/components/file-browser/FileBrowser.ts:
```diff
@@ -0,0 +1,54 @@
+          background: '#fff'
+        }
+      }, [
+        h(FileBrowserDirectories, {
```
I don't understand-- why is there both a `FileBrowserDirectories` and a `FileBrowserFiles` (down at line 49)? I am wondering if some documentation would help make it clearer how everything fits together.

nawatts replied:
FileBrowserDirectories is the left panel with the directory tree. FileBrowserFiles is the right panel with the files in the selected directory.

cahrens replied:
Ah, what was confusing me is that a directory includes files…. I thought FileBrowserFiles was like a child node inside of a directory.

nawatts replied:
I could rename to something like FileBrowserDirectoryTree and FileBrowserFilesInDirectory?

cahrens replied:
Maybe just `DirectoryTree` and `FilesInDirectory` or `DirectoryContents`? Everything is already in `file-browser`, so I wonder if if it's really useful to prefix every file with FileBrowser.

nawatts replied:
👍  I tend to name files for navigation via fuzzy finder. In other parts of the code, we have four files all named List.js.

cahrens replied:
I can see that, but here "FileBrowser" makes me think very specifically of the directory tree, which added to my confusion.

nawatts replied:
Renamed in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/7532cdb1ae394e50958754ce90ee32b56a542f76.

cahrens commented on src/components/file-browser/FileBrowserDirectories.test.ts:
```diff
@@ -0,0 +1,244 @@
+    expect(contentsFetchedBeforeExpanding).toBe(false)
+    expect(contentsFetchedAfterExpanding).toBe(true)
+
+    expect(renderedSubdirectories).toEqual(['subdirectory1', 'subdirectory2', 'subdirectory3'])
```
It might also be good to add in an axe check to verify aria attributes/ids. 

```
expect(await axe(container)).toHaveNoViolations()
```

nawatts replied:
Added in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/ca23519197059376718c5a5107f814af4b2ff00c.

cahrens commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,369 @@
+
+  const {
+    state: { status, directories },
+    hasNextPage,
```
Why are there no types on these?

nawatts replied:
They are known from the return type of `useDirectoriesInDirectory`.

cahrens replied:
This is something I'm trying to get used to… that the types aren't repeated so when you are reading the code outside of an IDE it's not easy to tell what the types are.

cahrens commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,369 @@
+  } = useDirectoriesInDirectory(provider, path)
+
+  useEffect(() => {
+    if (status === 'Ready' || status === 'Error') {
```
Should status be an enum?

nawatts replied:
We decided to use unions of literals instead of enums. `status` is type checked as one of "Loading", "Ready" or "Error".

cahrens replied:
Oh, can you point me to where the status literals are?

nawatts replied:
[src/libs/type-utils/LoadedState.ts](https://github.com/DataBiosphere/terra-ui/blob/su-229-file-browser-2/src/libs/type-utils/LoadedState.ts).

`useDirectoriesInDirectory` uses `useIncrementalResponse`. The type of the `state` property returned from `useIncrementalResponse` is based on `LoadedState`.

https://github.com/DataBiosphere/terra-ui/blob/1f75600e7cb8165e47a5391b876c1dedd47f2d87/src/components/file-browser/file-browser-hooks.ts#L16-L24

https://github.com/DataBiosphere/terra-ui/blob/1f75600e7cb8165e47a5391b876c1dedd47f2d87/src/libs/ajax/incremental-response/useIncrementalResponse.ts#L9-L17

cahrens replied:
Interesting… this will take me some time to get used to also!

cahrens commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,369 @@
+        [!isExpanded, () => 'angle-right']
+      ), {
+        // @ts-expect-error
+        color: isSelected ? '#000' : colors.accent(),
```
It seems like we should define `#000` as a variable (at least within this file).

nawatts replied:
Done in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/9ddd07e5eacaf47087d534a45b20cdc276230049.

nawatts committed:
Rename components

nawatts committed:
Define constant for selected directory color

cahrens commented on src/components/file-browser/FileBrowserDirectories.ts:
```diff
@@ -0,0 +1,369 @@
+        setActiveDescendant('node-0')
+      } else if (e.key === 'Enter') {
+        if (currentTreeItem.id.endsWith('-load-next')) {
+          // If on a load next page tree item, load the next page.
```
It would be nice if the keyboard focus went to the first loaded node when "Load Next Page" completes. As it is, you have to navigate back up and understand what was there before you triggered the action.

nawatts replied:
Added in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/2b09e10ee894418346ff07931134e02ba78f8a45.

nawatts committed:
Add axe to test

nawatts committed:
Update aria-activedescendant after loading next page
